sdk/core/azure-core/src/main/java/com/azure/core/util/QueryParameter.java

```diff
@@ -0,0 +1,144 @@
+     *
+     * @param name the name of the parameter.
+     * @param value the value of the parameter.
+     * @throws NullPointerException if {@code name} is null.
+     */
+    public QueryParameter(String name, String value) {
+        Objects.requireNonNull(name, "'name' cannot be null.");
+        this.name = name;
+        this.value = value;
+    }
+
+    /**
+     * Create a QueryParameter instance using the provided name and values.
+     *
+     * @param name the name of the parameter.
+     * @param values the values of the parameter.
+     * @throws NullPointerException if {@code name} or {@code values} are null.
+     */
+    public QueryParameter(String name, List<String> values) {
+        Objects.requireNonNull(name, "'name' cannot be null.");
+        Objects.requireNonNull(values, "'values' cannot be null");
+        this.name = name;
+        this.value = null;
+        this.values = new LinkedList<>(values);
+    }
+
+    /**
+     * Gets the parameter name.
+     *
+     * @return the name of this {@link QueryParameter}
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Gets the combined, comma-separated value of this {@link QueryParameter}, taking into account all values provided.
+     *
+     * @return the value of this QueryParameter
+     */
+    public String getValue() {
+        checkCachedStringValue();
+        return cachedStringValue;
+    }
+
+    /**
+     * Gets the comma separated value as an array. Changes made to this array will not be reflected in the parameters.
+     *
+     * @return the values of this {@link QueryParameter} that are separated by a comma
+     */
+    public String[] getValues() {
+        if (values == null) {
+            // most common case
+            return new String[] {value};
+        } else {
+            return values.toArray(new String[] { });
+        }
+    }
+
+    /**
+     * Returns all values associated with this parameter, represented as an unmodifiable list of strings.
+     *
+     * @return An unmodifiable list containing all values associated with this parameter.
+     */
+    public List<String> getValuesList() {
+        if (values == null) {
+            // most common case is that we don't have a list of values, but a single one
+            // a convenience return value is implemented here to avoid NPEs.
+            // List.of() would be a better option but it is Java 9+ only.
+            return Collections.unmodifiableList(Arrays.asList(value));
+        } else {
+            return Collections.unmodifiableList(values);
+        }
+    }
+
+    /**
+     * Add a new value to the end of the QueryParameter.
+     *
+     * @param value the value to add
+     */
+    public void addValue(String value) {
+        if (values == null) {
+            values = new LinkedList<>();
+            values.add(this.value); // add current standalone value
+        }
+
+        values.add(value);
```

Comment:
What you're doing here is right, but it so much looks like a bug that I bet someone will be tempted to 'fix' it. If you don't have a unit test, please do make sure to test this scenario. Even still, you might be tempted to rename the parameter to `newValue` or something, to differentiate from the `this.value`.

Reply:
Change done. This is covered by the unit tests.